File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(vector<string> friends, vector<string> gifts) 
{
    int n = friends.size();
    // Map each friend's name to a unique index
    unordered_map<string, int> name_to_index;
    for(int i = 0; i < n; ++i)
    {
        name_to_index[friends[i]] = i;
    }
    
    // Initialize arrays to track total given and received
    vector<int> given(n, 0);
    vector<int> received(n, 0);
    
    // Initialize a 2D array to track gifts given from A to B
    vector<vector<int>> gifts_given(n, vector<int>(n, 0));
    
    // Process each gift record
    for(auto &gift : gifts)
    {
        // Split the string into giver and receiver
        size_t space_pos = gift.find(' ');
        string A = gift.substr(0, space_pos);
        string B = gift.substr(space_pos + 1);
        
        int a = name_to_index[A];
        int b = name_to_index[B];
        
        given[a] += 1;
        received[b] += 1;
        gifts_given[a][b] += 1;
    }
    
    // Calculate gift index for each friend
    vector<int> gift_index(n, 0);
    for(int i = 0; i < n; ++i)
    {
        gift_index[i] = given[i] - received[i];
    }
    
    // Initialize array to track gifts each friend will receive next month
    vector<int> received_next(n, 0);
    
    // Iterate over all unique pairs of friends
    for(int i = 0; i < n; ++i){
        for(int j = i+1; j < n; ++j)
        {
            int a_to_b = gifts_given[i][j];
            int b_to_a = gifts_given[j][i];
            
            if(a_to_b > b_to_a)
            {
                // Friend i has given more to j, so i receives one gift from j
                received_next[i] += 1;
            }
            else if(b_to_a > a_to_b)
            {
                // Friend j has given more to i, so j receives one gift from i
                received_next[j] += 1;
            }
            else{
                // Equal number of gifts given to each other, compare gift indices
                if(gift_index[i] > gift_index[j])
                {
                    // Friend i has a higher gift index, receives one gift from j
                    received_next[i] += 1;
                }
                else if(gift_index[j] > gift_index[i])
                {
                    // Friend j has a higher gift index, receives one gift from i
                    received_next[j] += 1;
                }
                // If gift indices are equal, no one receives a gift
            }
        }
    }
    
    // Find the maximum number of gifts any friend will receive
    int max_gifts = 0;
    for(auto &count : received_next)
    {
        if(count > max_gifts)
        {
            max_gifts = count;
        }
    }
    
    return max_gifts;
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution 
{
    public int solution(string[] friends, string[] gifts) 
    {
        Dictionary<string, Dictionary<string,int>> dic = new Dictionary<string, Dictionary<string,int>>(); 
        Dictionary<string, int> giveDic = new Dictionary<string, int>();
        Dictionary<string, int> takeDic = new Dictionary<string, int>();
        // 딕셔너리 초기화 과정
        foreach(string friend in friends)
        {
            Dictionary<string, int> tmpDic = new Dictionary<string, int>();
            dic[friend] = tmpDic;
            giveDic[friend] = 0;
            takeDic[friend] = 0;
        }
        
        foreach(string gift in gifts)
        {
            string[] split = gift.Split(' ');
            string give = split[0];
            string take = split[1];
            
            if (!dic[give].ContainsKey(take))
            {
                dic[give][take] = 0;
            }
            dic[give][take]++;
    
            giveDic[give]++;
            takeDic[take]++;
        }
        
        Dictionary<string, int> giftVauleDic = new Dictionary<string, int>();
        foreach(string friend in friends)
        {
            giftVauleDic[friend] = giveDic[friend] - takeDic[friend];
        }
        
        int[] nextMonth = new int[friends.Length];  // 각 친구가 다음달에 받을 선물 수

        for (int i = 0; i < friends.Length; i++)
        {
            for (int j = i + 1; j < friends.Length; j++)  // i보다 뒤에 있는 것만 (중복 방지)
            {
                string a = friends[i];
                string b = friends[j];

                int aToB = dic[a].GetValueOrDefault(b, 0);  // a가 b에게 준 횟수
                int bToA = dic[b].GetValueOrDefault(a, 0);  // b가 a에게 준 횟수

                if (aToB > bToA)
                {
                    // a가 b에게 더 많이 줬으니, b가 a에게 하나 받음
                    nextMonth[i]++;
                }
                else if (aToB < bToA)
                {
                    nextMonth[j]++;
                }
                else
                {
                    // 주고받은 기록이 같거나 없으면 → 선물 지수 비교
                    if (giftVauleDic[a] > giftVauleDic[b])
                    {
                        nextMonth[i]++;
                    }
                    else if (giftVauleDic[a] < giftVauleDic[b])
                    {
                        nextMonth[j]++;
                    }
                    // 지수도 같으면 아무도 안 받음
                }
            }
        }

        return nextMonth.Max();
            
    }
}